homeassistant/components/mqtt/subscription.py

```diff
@@ -5,45 +5,75 @@
+class EntitySubscription:
+    """Class to hold data about an active entity topic subscription."""
+
+    topic = attr.ib(type=str)
+    message_callback = attr.ib(type=MessageCallbackType)
+    unsubscribe_callback = attr.ib(type=Optional[Callable[[], None]])
+    qos = attr.ib(type=int, default=0)
+    encoding = attr.ib(type=str, default='utf-8')
+
+    def should_resubscribe(self, other):
+        """Check if we should re-subscribe to the topic using the old state."""
+        if other is None:
+            return True
+
+        return self.topic != other.topic or self.qos != other.qos or \
+            self.encoding != other.encoding
+
+
 @bind_hass
-async def async_subscribe_topics(hass: HomeAssistantType, sub_state: dict,
-                                 topics: dict):
+async def async_subscribe_topics(hass: HomeAssistantType,
+                                 new_state: Optional[Dict[str,
+                                                          EntitySubscription]],
+                                 topics: Dict[str, Any]):
     """(Re)Subscribe to a set of MQTT topics.
 
-    State is kept in sub_state.
+    State is kept in sub_state and a dictionary mapping from the subscription
+    key to the subscription state
     """
-    cur_state = sub_state if sub_state is not None else {}
-    sub_state = {}
-    for key in topics:
-        topic = topics[key].get('topic', None)
-        msg_callback = topics[key].get('msg_callback', None)
-        qos = topics[key].get('qos', DEFAULT_QOS)
-        encoding = topics[key].get('encoding', 'utf-8')
-        topic = (topic, msg_callback, qos, encoding)
-        (cur_topic, unsub) = cur_state.pop(
-            key, ((None, None, None, None), None))
-
-        if topic != cur_topic and topic[0] is not None:
-            if unsub is not None:
-                unsub()
-            unsub = await mqtt.async_subscribe(
-                hass, topic[0], topic[1], topic[2], topic[3])
-        sub_state[key] = (topic, unsub)
-
-    for key, (topic, unsub) in list(cur_state.items()):
-        if unsub is not None:
-            unsub()
+    current_subscriptions = new_state if new_state is not None else {}
+    new_state = {}
+    for key, value in topics.items():
+        # Extract the new requested subscription
+        requested = EntitySubscription(
+            topic=value.get('topic', None),
+            message_callback=value.get('msg_callback', None),
+            unsubscribe_callback=None,
+            qos=value.get('qos', DEFAULT_QOS),
+            encoding=value.get('encoding', 'utf-8'),
+        )
+        # Get the current subscription state
+        current = current_subscriptions.pop(key, None)
 
-    return sub_state
+        # Re-subscribe if we need to
+        if requested.should_resubscribe(current):
+            if current.unsubscribe_callback is not None:
+                current.unsubscribe_callback()
+            requested.unsubscribe_callback = await mqtt.async_subscribe(
```

Comment:
Should this be moved into your class too ?

Reply:
Ah yes, let me do that.